File: cpp/leetcode/081.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int l = 0, r = nums.size() - 1;
        while(l <= r) {
            int m = l + (r - l) / 2;
            if(target == nums[m]) return true;
            if(nums[m] > nums[l]) {
                if(nums[l] <= target && target < nums[m]) r = m - 1;
                else l = m + 1;
            } else if(nums[m] < nums[l]) {
                if(nums[r] >= target && target > nums[m]) l = m + 1;
                else r = m - 1;
            } else {
                l ++;
            }
        }
        return false;
    }
};

File: python/leetcode/081.py
class Solution:
    def search(self, nums, target):
        l, r = 0, len(nums) - 1
        while l <= r:
            m = (r - l)/2 + l
            if nums[m] == target: return True
            if nums[m] > nums[l]:
                if nums[l] <= target < nums[m]: r = m - 1
                else: l = m + 1
            elif nums[m] < nums[l]:
                if nums[m] < target <= nums[r]: l = m + 1
                else: r = m - 1
            else:  l += 1
        return False